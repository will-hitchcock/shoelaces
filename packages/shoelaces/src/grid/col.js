import {
  merge, omitBy, isUndefined, flow, isNumber,
} from 'lodash/fp';
import flex from './flex';
import query from './media.new';

const getSize = (num) => `${num * 100}%`;

const getWidth = (size) => {
  if (!size) return {};
  if (size === 'auto') {
    return {
      flexGrow: 1,
      flexBasis: 0,
      maxWidth: '100%',
    };
  }

  const width = getSize(size);

  return {
    flexBasis: `${width}`,
    maxWidth: `${width}`,
  };
};

const getOffset = (num) => (isNumber(num) ? ({ marginLeft: `${getSize(num)}` }) : undefined);

const col = ({
  column = true, gutter = '0.5em', offset, size, xs, sm, md, lg, xl, xx, ...props
} = {}) => ({
  boxSizing: 'border-box',
  flex: '0 0 auto',
  padding: gutter,
  ...flex({ column, ...props }),
  ...merge(
    query(getOffset)(offset),
    query(getWidth)({
      xs, sm, md, lg, xl, xx,
    }),
  ),
});

export default flow(
  col,
  omitBy(isUndefined),
);
