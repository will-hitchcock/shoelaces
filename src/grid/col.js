import { flattenMin } from '../util/media'
import render from '../util/render'
/**
*
* Utility function for calculating col size
* Values must be between 0 and 1
*
**/
const colSize = num => num * 100 + '%'

/**
*
* Mixin for setting column width
* returns css for flex-basis and max-width depending on num
* uses colSize to determine width
* also accepts 'auto'
*
**/
const colWidth = size => {
  if (size === 'auto') {
    return {
      flexGrow: 1,
      flexBasis: 0,
      maxWidth: '100%'
    }
  }

  const width = colSize(size)

  return {
    flexBasis: `${width}`,
    maxWidth: `${width}`
  }
}

/**
*
* Mixin for setting offset
* uses colSize to calculate margin for the given side
*
**/
// TODO: Fix the colOffset, currently the prog only accepts a number
// this won't work in responsive column implementations
const colOffset = num => {
  if (!num) return {}
  return {
    marginLeft: `${colSize(num)}`
  }
}

const col = ({ gutter, reverse, offset, ...props }) => ({
  boxSizing: 'border-box',
  flex: '0 0 auto',
  padding: gutter,
  flexDirection: reverse ? 'column-reverse' : 'initial',
  ...colOffset(offset),
  ...flattenMin(props, colWidth)
})

col.defaultProps = {
  gutter: '0.5em'
}

export default render(col)
