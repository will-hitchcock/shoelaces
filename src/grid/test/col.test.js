import col from '../col'

describe('col', () => {
  it('renders default col styles', () => {
    expect(col()).toMatchSnapshot()
  })

  it('renders reverse col styles', () => {
    expect(col({ reverse: true })).toMatchSnapshot()
  })

  it('renders 0.5 offset col styles', () => {
    expect(col({ offset: 0.5 })).toMatchSnapshot()
  })

  it('renders size xs 0.5 col styles', () => {
    expect(col({ xs: 0.5 })).toMatchSnapshot()
  })

  it('renders size md auto col styles', () => {
    expect(col({ md: 'auto' })).toMatchSnapshot()
  })
})
