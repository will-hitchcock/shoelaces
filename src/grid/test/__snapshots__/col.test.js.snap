// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`col renders 0.5 offset col styles 1`] = `
Object {
  "alignContent": "flex-start",
  "alignItems": "stretch",
  "alignSelf": "auto",
  "boxSizing": "border-box",
  "display": "flex",
  "flex": "0 0 auto",
  "flexDirection": "column",
  "justifyContent": "space-between",
  "marginLeft": "50%",
  "padding": "0.5em",
}
`;

exports[`col renders 0.5 xs offset col styles 1`] = `
Object {
  "alignContent": "flex-start",
  "alignItems": "stretch",
  "alignSelf": "auto",
  "boxSizing": "border-box",
  "display": "flex",
  "flex": "0 0 auto",
  "flexDirection": "column",
  "justifyContent": "space-between",
  "marginLeft": "50%",
  "padding": "0.5em",
}
`;

exports[`col renders default col styles 1`] = `
Object {
  "alignContent": "flex-start",
  "alignItems": "stretch",
  "alignSelf": "auto",
  "boxSizing": "border-box",
  "display": "flex",
  "flex": "0 0 auto",
  "flexDirection": "column",
  "justifyContent": "space-between",
  "padding": "0.5em",
}
`;

exports[`col renders reverse col styles 1`] = `
Object {
  "alignContent": "flex-start",
  "alignItems": "stretch",
  "alignSelf": "auto",
  "boxSizing": "border-box",
  "display": "flex",
  "flex": "0 0 auto",
  "flexDirection": "column-reverse",
  "justifyContent": "space-between",
  "padding": "0.5em",
}
`;

exports[`col renders size 0.5 and offset 0.5 width at sm and size 1 offset 0 at xs 1`] = `
Object {
  "@media (min-width: 36em)": Object {
    "flexBasis": "50%",
    "marginLeft": "50%",
    "maxWidth": "50%",
  },
  "alignContent": "flex-start",
  "alignItems": "stretch",
  "alignSelf": "auto",
  "boxSizing": "border-box",
  "display": "flex",
  "flex": "0 0 auto",
  "flexBasis": "100%",
  "flexDirection": "column",
  "justifyContent": "space-between",
  "marginLeft": "0%",
  "maxWidth": "100%",
  "padding": "0.5em",
}
`;

exports[`col renders size md auto col styles 1`] = `
Object {
  "@media (min-width: 48em)": Object {
    "flexBasis": 0,
    "flexGrow": 1,
    "maxWidth": "100%",
  },
  "alignContent": "flex-start",
  "alignItems": "stretch",
  "alignSelf": "auto",
  "boxSizing": "border-box",
  "display": "flex",
  "flex": "0 0 auto",
  "flexDirection": "column",
  "justifyContent": "space-between",
  "padding": "0.5em",
}
`;

exports[`col renders size xs 0.5 col styles 1`] = `
Object {
  "alignContent": "flex-start",
  "alignItems": "stretch",
  "alignSelf": "auto",
  "boxSizing": "border-box",
  "display": "flex",
  "flex": "0 0 auto",
  "flexBasis": "50%",
  "flexDirection": "column",
  "justifyContent": "space-between",
  "maxWidth": "50%",
  "padding": "0.5em",
}
`;
